File: snake.js
/**
 * 贪吃蛇游戏 - 带迷宫墙壁的增强版
 * 游戏特点：
 * 1. 动态迷宫生成
 * 2. 多食物系统
 * 3. 关卡进阶机制
 */

// 获取DOM元素
const canvas = document.getElementById('game-canvas');
const ctx = canvas.getContext('2d');
const scoreElement = document.getElementById('score');
const gameOverElement = document.getElementById('game-over');
const finalScoreElement = document.getElementById('final-score');
const restartButton = document.getElementById('restart-btn');

// 游戏基本配置
const gridSize = 20; // 网格大小，单位像素
const tileCount = canvas.width / gridSize; // 网格数量

// 游戏状态变量
let snake = [
    { x: 10, y: 10 } // 蛇的初始位置
];
let foods = [{ x: 15, y: 15 }]; // 食物数组
let walls = []; // 墙壁数组
let dx = 0; // 水平移动方向
let dy = 0; // 垂直移动方向
let score = 0; // 分数
let gameLoop; // 游戏循环计时器
let isGeneratingNewMaze = false; // 是否正在生成新迷宫的标志

/**
 * 游戏主循环函数
 * 负责清屏、移动蛇、检测碰撞、绘制游戏元素和更新分数
 */
function drawGame() {
    clearCanvas();
    moveSnake();
    checkCollision();
    drawWalls(); // 绘制墙壁
    drawFood();
    drawSnake();
    updateScore();
}

/**
 * 清空画布
 */
function clearCanvas() {
    ctx.fillStyle = '#000';
    ctx.fillRect(0, 0, canvas.width, canvas.height);
}

/**
 * 绘制蛇
 * 使用绿色填充，每个蛇身体段略小于网格大小，形成网格间隙效果
 */
function drawSnake() {
    ctx.fillStyle = '#4CAF50';
    snake.forEach(segment => {
        ctx.fillRect(segment.x * gridSize, segment.y * gridSize, gridSize - 2, gridSize - 2);
    });
}

/**
 * 绘制食物
 * 使用红色填充，每个食物略小于网格大小
 */
function drawFood() {
    ctx.fillStyle = '#ff0000';
    foods.forEach(food => {
        ctx.fillRect(food.x * gridSize, food.y * gridSize, gridSize - 2, gridSize - 2);
    });
}

/**
 * 绘制墙壁
 * 使用蓝色填充，墙壁填满整个网格
 */
function drawWalls() {
    ctx.fillStyle = '#3333cc';
    walls.forEach(wall => {
        ctx.fillRect(wall.x * gridSize, wall.y * gridSize, gridSize, gridSize);
    });
}

/**
 * 生成迷宫墙壁
 * 算法特点：
 * 1. 保留部分旧墙壁，增加新墙壁，保持游戏连续性
 * 2. 边界墙壁和内部墙壁分开生成，确保迷宫结构合理
 * 3. 确保墙壁不会太靠近蛇头，避免突然死亡
 */
function generateMaze() {
    const oldWalls = [...walls]; // 保存当前的墙壁
    const totalWallCount = oldWalls.length; // 记录当前墙壁总数
    walls = [];
    
    // 获取蛇头位置，用于确保墙壁不会太靠近蛇头
    const snakeHead = snake[0];
    
    // 如果有现有的墙壁，保留90%
    if (oldWalls.length > 0) {
        // 随机选择90%的墙壁保留
        const wallsToKeep = Math.floor(oldWalls.length * 0.9);
        // 随机打乱墙壁数组，确保随机性
        const shuffledWalls = oldWalls.sort(() => Math.random() - 0.5);
        walls = shuffledWalls.slice(0, wallsToKeep);
    }
    
    // 计算需要新生成的墙壁数量，确保总数保持相对稳定
    let wallsToGenerate = 0;
    if (totalWallCount > 0) {
        // 如果有现有墙壁，只需要生成10%的新墙壁
        wallsToGenerate = totalWallCount - walls.length;
    } else {
        // 如果是首次生成，则创建一个基本数量的墙壁
        // 边界墙壁数量估计(周长的20%) + 内部墙壁数量(面积的2%)
        wallsToGenerate = Math.floor(tileCount * 4 * 0.2) + Math.floor(tileCount * tileCount * 0.02);
    }
    
    // 首先尝试在边界生成一些墙壁
    let boundaryWallsAdded = 0;
    // 最多70%的新墙用于边界，确保内部也有足够的墙壁
    const maxBoundaryWalls = Math.min(wallsToGenerate, Math.floor(wallsToGenerate * 0.7)); 
    
    for (let i = 0; i < tileCount && boundaryWallsAdded < maxBoundaryWalls; i++) {
        // 上边界和下边界
        if (i < 3 || i > tileCount - 4) {
            continue; // 留出边缘空间，避免游戏开始就结束
        }
        
        // 在四个边界上定义潜在的墙壁位置
        const topWall = { x: i, y: 5 };
        const bottomWall = { x: i, y: tileCount - 6 };
        const leftWall = { x: 5, y: i };
        const rightWall = { x: tileCount - 6, y: i };
        
        // 计算每个潜在墙壁与蛇头的曼哈顿距离
        const distanceToTopWall = Math.abs(topWall.x - snakeHead.x) + Math.abs(topWall.y - snakeHead.y);
        const distanceToBottomWall = Math.abs(bottomWall.x - snakeHead.x) + Math.abs(bottomWall.y - snakeHead.y);
        const distanceToLeftWall = Math.abs(leftWall.x - snakeHead.x) + Math.abs(leftWall.y - snakeHead.y);
        const distanceToRightWall = Math.abs(rightWall.x - snakeHead.x) + Math.abs(rightWall.y - snakeHead.y);
        
        // 只有当墙与蛇头距离足够远时才添加，并且控制添加的数量
        // 10%的概率添加墙壁，确保墙壁分布均匀但不过密
        if (Math.random() < 0.1 && distanceToTopWall >= 3 && boundaryWallsAdded < maxBoundaryWalls) {
            walls.push(topWall);
            boundaryWallsAdded++;
        }
        
        if (Math.random() < 0.1 && distanceToBottomWall >= 3 && boundaryWallsAdded < maxBoundaryWalls) {
            walls.push(bottomWall);
            boundaryWallsAdded++;
        }
        
        if (Math.random() < 0.1 && distanceToLeftWall >= 3 && boundaryWallsAdded < maxBoundaryWalls) {
            walls.push(leftWall);
            boundaryWallsAdded++;
        }
        
        if (Math.random() < 0.1 && distanceToRightWall >= 3 && boundaryWallsAdded < maxBoundaryWalls) {
            walls.push(rightWall);
            boundaryWallsAdded++;
        }
    }
    
    // 生成剩余的内部随机墙壁
    const remainingWallsToGenerate = wallsToGenerate - boundaryWallsAdded;
    for (let i = 0; i < remainingWallsToGenerate; i++) {
        generateSingleWall();
    }
}

/**
 * 生成单个墙壁
 * 确保墙壁不会生成在蛇身上、食物上，或离蛇头太近
 * 同时确保墙壁之间有足够的空间，避免形成无法通过的区域
 */
function generateSingleWall() {
    // 在游戏区域中心区域生成墙壁，避开边缘5格
    const newWall = {
        x: Math.floor(Math.random() * (tileCount - 10)) + 5,
        y: Math.floor(Math.random() * (tileCount - 10)) + 5
    };
    
    // 获取蛇头位置
    const snakeHead = snake[0];
    // 计算与蛇头的曼哈顿距离
    const distanceToHead = Math.abs(newWall.x - snakeHead.x) + Math.abs(newWall.y - snakeHead.y);
    
    // 检查墙壁是否会与现有元素重叠或太靠近蛇头
    if (snake.some(segment => segment.x === newWall.x && segment.y === newWall.y) ||
        foods.some(food => food.x === newWall.x && food.y === newWall.y) ||
        walls.some(wall => wall.x === newWall.x && wall.y === newWall.y) ||
        distanceToHead < 3) { // 确保与蛇头至少保持3个格子的距离
        // 如果位置不合适，递归尝试生成新位置
        generateSingleWall();
        return;
    }
    
    // 检查墙壁周围是否有足够的空间，避免形成无法通过的区域
    let hasNearbyWall = false;
    for (let wall of walls) {
        const distance = Math.abs(wall.x - newWall.x) + Math.abs(wall.y - newWall.y);
        if (distance < 2) { // 如果有相邻的墙，不添加这个墙
            hasNearbyWall = true;
            break;
        }
    }
    
    // 只有当周围没有其他墙壁时，才添加这个新墙壁
    if (!hasNearbyWall) {
        walls.push(newWall);
    } else {
        // 如果位置不合适，递归尝试生成新位置
        generateSingleWall();
    }
}

/**
 * 移动蛇
 * 处理蛇的移动、食物的吃取和关卡进阶
 */
function moveSnake() {
    // 根据当前方向创建新的蛇头
    const head = { x: snake[0].x + dx, y: snake[0].y + dy };
    // 将新蛇头添加到蛇身体的前端
    snake.unshift(head);
    
    // 检查是否吃到任何一个食物
    let foodEaten = false;
    for (let i = 0; i < foods.length; i++) {
        if (head.x === foods[i].x && head.y === foods[i].y) {
            score += 10; // 每吃一个食物得10分
            foods.splice(i, 1); // 移除被吃掉的食物
            foodEaten = true;
            break;
        }
    }
    
    // 关卡进阶：如果没有食物了，重新生成迷宫和食物
    if (foods.length === 0 && !isGeneratingNewMaze) {
        isGeneratingNewMaze = true;
        setTimeout(() => {
            generateMaze(); // 重新生成迷宫
            generateFood(); // 生成新的食物
            isGeneratingNewMaze = false;
        }, 1000); // 1秒的延迟，给玩家一个视觉提示
    }
    
    // 如果没有吃到食物，移除蛇尾，保持蛇长度不变
    if (!foodEaten) {
        snake.pop();
    }
    // 如果吃到食物，不移除蛇尾，蛇长度+1
}

/**
 * 生成食物
 * 随机生成1-3个食物，增加游戏的挑战性和策略性
 */
function generateFood() {
    // 清空现有食物数组
    foods = [];
    
    // 随机生成1-3个食物
    const foodCount = Math.floor(Math.random() * 3) + 1; // 1-3个食物
    
    // 为每个食物调用单独的生成函数
    for (let i = 0; i < foodCount; i++) {
        generateSingleFood();
    }
}

/**
 * 生成单个食物
 * 确保食物不会生成在蛇身上、其他食物上或墙壁上
 */
function generateSingleFood() {
    const newFood = {
        x: Math.floor(Math.random() * tileCount),
        y: Math.floor(Math.random() * tileCount)
    };
    
    // 确保食物不会生成在蛇身上、其他食物上或墙壁上
    if (snake.some(segment => segment.x === newFood.x && segment.y === newFood.y) ||
        foods.some(food => food.x === newFood.x && food.y === newFood.y) ||
        walls.some(wall => wall.x === newFood.x && wall.y === newFood.y)) {
        generateSingleFood();
    } else {
        foods.push(newFood);
    }
}

/**
 * 检查碰撞
 * 检测蛇头是否碰到游戏边界、迷宫墙壁或自己的身体
 */
function checkCollision() {
    const head = snake[0];
    
    // 检查是否撞墙（游戏边界）
    if (head.x < 0 || head.x >= tileCount || head.y < 0 || head.y >= tileCount) {
        gameOver();
    }
    
    // 检查是否撞到迷宫墙壁
    if (walls.some(wall => wall.x === head.x && wall.y === head.y)) {
        gameOver();
    }
    
    // 检查是否撞到自己
    for (let i = 1; i < snake.length; i++) {
        if (head.x === snake[i].x && head.y === snake[i].y) {
            gameOver();
        }
    }
}

/**
 * 更新分数显示
 * 将当前分数更新到页面上的分数元素中
 */
function updateScore() {
    scoreElement.textContent = `分数: ${score}`;
}

/**
 * 游戏结束处理
 * 停止游戏循环，显示最终分数和游戏结束界面
 */
function gameOver() {
    clearInterval(gameLoop);
    finalScoreElement.textContent = `最终分数: ${score}`;
    gameOverElement.style.display = 'block';
}

/**
 * 重置游戏
 * 将所有游戏状态恢复到初始值，隐藏游戏结束界面，重新生成迷宫和食物，启动游戏
 */
function resetGame() {
    snake = [{ x: 10, y: 10 }];
    foods = [];
    walls = [];
    dx = 0;
    dy = 0;
    score = 0;
    gameOverElement.style.display = 'none';
    generateMaze(); // 生成迷宫
    generateFood(); // 生成初始食物
    startGame();
}

/**
 * 处理键盘按键事件
 * 根据用户按下的方向键改变蛇的移动方向
 * 包含防止180度掉头的逻辑，避免蛇立即撞到自己
 * @param {KeyboardEvent} e - 键盘事件对象
 */
function handleKeyPress(e) {
    const key = e.key;
    
    // 检测R键重启游戏
    if (key === 'r' || key === 'R') {
        resetGame();
        return;
    }
    
    // 防止180度掉头
    if (key === 'ArrowUp' && dy !== 1) { // 不允许向上时向下掉头
        dx = 0;
        dy = -1;
    } else if (key === 'ArrowDown' && dy !== -1) { // 不允许向下时向上掉头
        dx = 0;
        dy = 1;
    } else if (key === 'ArrowLeft' && dx !== 1) { // 不允许向左时向右掉头
        dx = -1;
        dy = 0;
    } else if (key === 'ArrowRight' && dx !== -1) { // 不允许向右时向左掉头
        dx = 1;
        dy = 0;
    }
}

/**
 * 启动游戏
 * 清除可能存在的旧游戏循环，设置新的游戏循环定时器
 * 游戏速度为每200毫秒更新一次
 */
function startGame() {
    if (gameLoop) {
        clearInterval(gameLoop);
    }
    gameLoop = setInterval(drawGame, 200);
}

// 设置事件监听器
document.addEventListener('keydown', handleKeyPress);
restartButton.addEventListener('click', resetGame);

// 初始化游戏
generateMaze(); // 首次生成迷宫
generateFood(); // 生成初始食物
startGame(); // 启动游戏循环

File: snake.py
import pygame
import random
import sys
from typing import List, Tuple

# 初始化Pygame
pygame.init()

# 游戏常量
WINDOW_SIZE = 1200
GRID_SIZE = 40
TILE_COUNT = WINDOW_SIZE // GRID_SIZE

# 颜色定义
BLACK = (0, 0, 0)
WHITE = (255, 255, 255)
GREEN = (76, 175, 80)
RED = (255, 0, 0)
BLUE = (51, 51, 204)

# 游戏状态
class GameState:
    def __init__(self):
        self.screen = pygame.display.set_mode((WINDOW_SIZE, WINDOW_SIZE))
        pygame.display.set_caption('贪吃蛇游戏')
        self.clock = pygame.time.Clock()
        self.font = pygame.font.SysFont('SimHei', 36)  # 使用黑体显示中文
        self.restart_button = None  # 重启按钮区域
        self.pause_button = None  # 暂停按钮区域
        self.reset()

    def reset(self):
        self.snake = [(10, 10)]  # 蛇的初始位置
        self.foods = [(15, 15)]  # 食物位置
        self.walls = []  # 墙壁位置
        self.direction = (0, 0)  # 初始方向设为静止
        self.score = 0
        self.game_over = False
        self.is_paused = False  # 添加暂停状态标志
        self.is_generating_maze = False
        self.move_counter = 0  # 添加移动计数器
        self.move_delay = 15    # 每隔6帧移动一次蛇
        pygame.time.set_timer(pygame.USEREVENT, 0)  # 确保停止任何正在进行的计时器
        self.restart_button = None  # 重置重启按钮
        self.pause_button = None  # 重置暂停按钮
        self.generate_maze()
        self.generate_food()

    def generate_single_wall(self):
        # 在游戏区域中心区域生成墙壁，避开边缘5格
        new_wall = (
            random.randint(5, TILE_COUNT - 6),
            random.randint(5, TILE_COUNT - 6)
        )

        # 获取蛇头位置
        snake_head = self.snake[0]
        # 计算与蛇头的曼哈顿距离
        distance_to_head = abs(new_wall[0] - snake_head[0]) + abs(new_wall[1] - snake_head[1])

        # 检查墙壁是否会与现有元素重叠或太靠近蛇头
        if (new_wall in self.snake or
            new_wall in self.foods or
            new_wall in self.walls or
            distance_to_head < 3):  # 确保与蛇头至少保持3个格子的距离
            return self.generate_single_wall()

        # 检查墙壁周围是否有足够的空间
        for wall in self.walls:
            distance = abs(wall[0] - new_wall[0]) + abs(wall[1] - new_wall[1])
            if distance < 2:  # 如果有相邻的墙，不添加这个墙
                return self.generate_single_wall()

        return new_wall

    def generate_maze(self):
        old_walls = self.walls.copy()
        total_wall_count = len(old_walls)
        self.walls = []

        # 如果有现有的墙壁，保留90%
        if old_walls:
            walls_to_keep = int(len(old_walls) * 0.9)
            random.shuffle(old_walls)
            self.walls = old_walls[:walls_to_keep]

        # 计算需要新生成的墙壁数量
        if total_wall_count > 0:
            walls_to_generate = total_wall_count - len(self.walls)
        else:
            # 首次生成：边界墙壁(周长20%) + 内部墙壁(面积2%)
            walls_to_generate = int(TILE_COUNT * 4 * 0.2) + int(TILE_COUNT * TILE_COUNT * 0.02)

        # 生成新的墙壁
        for _ in range(walls_to_generate):
            new_wall = self.generate_single_wall()
            if new_wall:
                self.walls.append(new_wall)

    def generate_food(self):
        self.foods = []
        food_count = random.randint(1, 3)
        for _ in range(food_count):
            self.generate_single_food()

    def generate_single_food(self):
        while True:
            new_food = (
                random.randint(0, TILE_COUNT - 1),
                random.randint(0, TILE_COUNT - 1)
            )
            if (new_food not in self.snake and
                new_food not in self.foods and
                new_food not in self.walls):
                self.foods.append(new_food)
                break

    def move_snake(self):
        # 如果游戏暂停，不执行移动逻辑
        if self.is_paused:
            return
            
        # 增加计数器
        self.move_counter += 1
        
        # 只有当计数器达到延迟值时才移动蛇
        if self.move_counter < self.move_delay:
            return
            
        # 重置计数器
        self.move_counter = 0
        
        if self.direction == (0, 0) or self.game_over:
            return
        
        # 计算新的蛇头位置
        new_head = (
            self.snake[0][0] + self.direction[0],
            self.snake[0][1] + self.direction[1]
        )

        # 检查碰撞
        if (new_head[0] < 0 or new_head[0] >= TILE_COUNT or
            new_head[1] < 0 or new_head[1] >= TILE_COUNT or
            new_head in self.snake or
            new_head in self.walls):
            self.game_over = True
            return

        # 移动蛇
        self.snake.insert(0, new_head)

        # 检查是否吃到食物
        if new_head in self.foods:
            self.score += 10
            self.foods.remove(new_head)
            if not self.foods and not self.is_generating_maze:
                self.is_generating_maze = True
                pygame.time.set_timer(pygame.USEREVENT, 1000)  # 1秒后生成新迷宫
        else:
            self.snake.pop()
            
    def check_button_click(self, pos):
        """检查是否点击了重启或暂停按钮"""
        if self.game_over and self.restart_button and self.restart_button.collidepoint(pos):
            return "restart"
        elif not self.game_over and self.pause_button and self.pause_button.collidepoint(pos):
            return "pause"
        return None

    def draw(self):
        self.screen.fill(BLACK)

        # 绘制墙壁
        for wall in self.walls:
            pygame.draw.rect(self.screen, BLUE,
                           (wall[0] * GRID_SIZE, wall[1] * GRID_SIZE,
                            GRID_SIZE, GRID_SIZE))

        # 绘制食物
        for food in self.foods:
            pygame.draw.rect(self.screen, RED,
                           (food[0] * GRID_SIZE, food[1] * GRID_SIZE,
                            GRID_SIZE - 2, GRID_SIZE - 2))

        # 绘制蛇
        for segment in self.snake:
            pygame.draw.rect(self.screen, GREEN,
                           (segment[0] * GRID_SIZE, segment[1] * GRID_SIZE,
                            GRID_SIZE - 2, GRID_SIZE - 2))

        # 绘制分数
        score_text = self.font.render(f'分数: {self.score}', True, WHITE)
        self.screen.blit(score_text, (10, 10))

        # 绘制暂停按钮（仅在游戏未结束时显示）
        if not self.game_over:
            pause_text = self.font.render('暂停' if not self.is_paused else '继续', True, WHITE)
            button_width = pause_text.get_width() + 40
            button_height = pause_text.get_height() + 20
            button_x = WINDOW_SIZE - button_width - 10
            button_y = 10
            
            # 绘制按钮背景
            pygame.draw.rect(self.screen, GREEN if not self.is_paused else RED,
                           (button_x, button_y, button_width, button_height), 0, 10)
            
            # 绘制按钮文字
            self.screen.blit(pause_text,
                           (button_x + 20, button_y + 10))
            
            # 保存按钮区域用于检测点击
            self.pause_button = pygame.Rect(button_x, button_y, button_width, button_height)

        # 如果游戏结束，显示游戏结束信息
        if self.game_over:
            game_over_text = self.font.render('游戏结束!', True, WHITE)
            final_score_text = self.font.render(f'最终分数: {self.score}', True, WHITE)
            restart_text = self.font.render('重新开始', True, WHITE)

            text_x = WINDOW_SIZE // 2
            self.screen.blit(game_over_text,
                           (text_x - game_over_text.get_width() // 2, WINDOW_SIZE // 2 - 60))
            self.screen.blit(final_score_text,
                           (text_x - final_score_text.get_width() // 2, WINDOW_SIZE // 2))
            
            # 绘制重启按钮
            button_width = restart_text.get_width() + 40
            button_height = restart_text.get_height() + 20
            button_x = text_x - button_width // 2
            button_y = WINDOW_SIZE // 2 + 60
            
            # 绘制按钮背景
            pygame.draw.rect(self.screen, GREEN, 
                           (button_x, button_y, button_width, button_height), 0, 10)
            
            # 绘制按钮文字
            self.screen.blit(restart_text,
                           (text_x - restart_text.get_width() // 2, button_y + 10))
            
            # 保存按钮区域用于检测点击
            self.restart_button = pygame.Rect(button_x, button_y, button_width, button_height)

        pygame.display.flip()

def main():
    game = GameState()
    running = True

    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
            elif event.type == pygame.MOUSEBUTTONDOWN:
                if event.button == 1:  # 左键点击
                    button_action = game.check_button_click(event.pos)
                    if button_action == "restart":
                        game.reset()
                        continue
                    elif button_action == "pause":
                        game.is_paused = not game.is_paused
                        continue
            elif event.type == pygame.KEYDOWN:
                if not game.game_over:
                    if event.key == pygame.K_SPACE:
                        # 空格键也可以暂停/恢复游戏
                        game.is_paused = not game.is_paused
                    elif not game.is_paused:  # 只有在游戏未暂停时才处理方向键
                        if event.key == pygame.K_UP and game.direction[1] != 1:
                            game.direction = (0, -1)
                        elif event.key == pygame.K_DOWN and game.direction[1] != -1:
                            game.direction = (0, 1)
                        elif event.key == pygame.K_LEFT and game.direction[0] != 1:
                            game.direction = (-1, 0)
                        elif event.key == pygame.K_RIGHT and game.direction[0] != -1:
                            game.direction = (1, 0)
            elif event.type == pygame.USEREVENT and game.is_generating_maze:
                game.generate_maze()
                game.generate_food()
                game.is_generating_maze = False
                pygame.time.set_timer(pygame.USEREVENT, 0)  # 停止计时器

        game.move_snake()
        game.draw()
        game.clock.tick(60)  # 提高帧率到60，使方向改变更快响应

    pygame.quit()
    sys.exit()

if __name__ == '__main__':
    main()